gdb/doc/gdb-add-index.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "GDB-ADD-INDEX 1"
.TH GDB-ADD-INDEX 1 2026-05-10 gdb-17.2 "GNU Development Tools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
gdb\-add\-index \- Add index files to speed up GDB
.SH SYNOPSIS
.IX Header "SYNOPSIS"
gdb-add-index [\-h | \-\-help] [\-v | \-\-version] [\-\-dwarf\-5] \fIfilename\fR
.SH DESCRIPTION
.IX Header "DESCRIPTION"
When GDB finds a symbol file, it scans the symbols in the
file in order to construct an internal symbol table.  This lets most
GDB operations work quickly\-\-at the cost of a delay early on.
For large programs, this delay can be quite lengthy, so GDB
provides a way to build an index, which speeds up startup.
.PP
To determine whether a file contains such an index, use the command
\&\f(CW\*(C`readelf \-S filename\*(C'\fR: the index is stored in a section named
\&\f(CW\*(C`.gdb_index\*(C'\fR (pre-DWARF 5) or \f(CW\*(C`.debug_names\*(C'\fR and
\&\f(CW\*(C`.debug_str\*(C'\fR (DWARF 5).  Indexes can only be produced on systems
which use ELF binaries and DWARF debug information (i.e., sections
named \f(CW\*(C`.debug_*\*(C'\fR).
.PP
By default \fBgdb-add-index\fR will add a pre-DWARF 5
\&\f(CW\*(C`.gdb_index\*(C'\fR section to \fIfilename\fR.  With \fB\-\-dwarf\-5\fR
the DWARF 5 section \f(CW\*(C`.debug_names\*(C'\fR is added instead.
.PP
\&\fIfilename\fR must be writable.
.PP
\&\fBgdb-add-index\fR uses GDB, \fBobjcopy\fR, and
\&\fBreadelf\fR found in the \fBPATH\fR environment variable.  If
you want to use different versions of these programs, you can specify
them through the appropriate environment variables (see below).
.PP
\&\fBgdb-add-index\fR exits with status 0 if it succeeds in creating
the index for \fIfilename\fR or greater than 0 if an error occurs.
.PP
See more in
the GDB manual in node \f(CW\*(C`Index Files\*(C'\fR
\&\-\- shell command \f(CW\*(C`info \-f gdb \-n "Index Files"\*(C'\fR.
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-\-dwarf\-5\fR 4
.IX Item "--dwarf-5"
Add DWARF 5 sections instead of previous \f(CW\*(C`.gdb_index\*(C'\fR section.
.IP \fB\-\-help\fR 4
.IX Item "--help"
.PD 0
.IP \fB\-h\fR 4
.IX Item "-h"
.PD
List all options, with brief explanations.
.IP \fB\-\-version\fR 4
.IX Item "--version"
.PD 0
.IP \fB\-v\fR 4
.IX Item "-v"
.PD
Print version information and then exit.
.SH ENVIRONMENT
.IX Header "ENVIRONMENT"
.IP \fBGDB\fR 4
.IX Item "GDB"
Full file name of the \f(CW\*(C`gdb\*(C'\fR program to use for index generation.
If not set, the \fBPATH\fR will be searched for a \f(CW\*(C`gdb\*(C'\fR program.
.IP \fBOBJCOPY\fR 4
.IX Item "OBJCOPY"
Full file name of the \f(CW\*(C`objcopy\*(C'\fR program to use to copy section
information into the given file.  If not set, the \fBPATH\fR will be searched
for a \f(CW\*(C`objcopy\*(C'\fR program.
.IP \fBREADELF\fR 4
.IX Item "READELF"
Full file name of the \f(CW\*(C`readelf\*(C'\fR program to use to inspect
properties of the given file.  If not set, the \fBPATH\fR will be searched
for a \f(CW\*(C`readelf\*(C'\fR program.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
The full documentation for GDB is maintained as a Texinfo manual.
If the \f(CW\*(C`info\*(C'\fR and \f(CW\*(C`gdb\*(C'\fR programs and GDB's Texinfo
documentation are properly installed at your site, the command
.PP
.Vb 1
\&        info gdb
.Ve
.PP
should give you access to the complete manual.
.PP
\&\fIUsing GDB: A Guide to the GNU Source-Level Debugger\fR,
Richard M. Stallman and Roland H. Pesch, July 1991.
.SH COPYRIGHT
.IX Header "COPYRIGHT"
Copyright (c) 1988\-2025 Free Software Foundation, Inc.
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.3 or
any later version published by the Free Software Foundation; with the
Invariant Sections being "Free Software" and "Free Software Needs
Free Documentation", with the Front-Cover Texts being "A GNU Manual,"
and with the Back-Cover Texts as in (a) below.
.PP
(a) The FSF's Back-Cover Text is: "You are free to copy and modify
this GNU Manual.  Buying copies from GNU Press supports the FSF in
developing GNU and promoting software freedom."


gdb/NEWS
		What has changed in GDB?
	     (Organized release by release)

*** Changes in GDB 17

* Debugging Linux programs that use x86-64 or x86-64 with 32-bit pointer
  size (X32) Shadow Stacks are now supported.

* Support for the shadow stack pointer register on x86-64 or x86-64 with
  32-bit pointer size (X32) GNU/Linux.

* Debugger Adapter Protocol changes

  ** GDB now supports the "completions" request.

  ** The attach request now accepts the coreFile parameter.

* "set style" commands now supports numeric format for basic colors
  from 0 to 255 and #RRGGBB format for TrueColor.

* New built-in convenience variable $_colorsupport provides comma-separated
  list of color space names supported by terminal.  Each color space name is one
  of monochrome, ansi_8color, aixterm_16color, xterm_256color or rgb_24bit.
  It is handy for conditionally using styling colors based on terminal features.
  For example:

  (gdb) if $_regex ($_colorsupport, ".*(^|,)rgb_24bit($|,).*")
   >set style filename background #FACADE
   >else
   >if $_regex ($_colorsupport, ".*(^|,)xterm_256color($|,).*")
    >set style filename background 224
    >else
    >set style filename background red
    >end
   >end

* UST (static tracepoint) support from gdbserver has been removed.

* Linux checkpoint code has been updated to work with multiple inferiors.

* The gcore and gdb-add-index scripts now have a -v or --version
  option, which prints the version number, and then exits.  As well as
  a -h or --help option, which prints each options and a brief
  description.

* On systems that support linker namespaces, the output of the command
  "info sharedlibraries" may add one more column, NS, which identifies the
  namespace into which the library was loaded, if more than one namespace
  is active.

* New built-in convenience variables $linker_namespace_count and
  $_linker_namespace.  These show the number of active linker
  namespaces, and the namespace to which the current location belongs to.
  In systems that don't support linker namespaces, or if the inferior hasn't
  started yet, these always return the integer 0.

* Add record full support for rv64gc architectures

* Debugging Linux programs that use AArch64 Guarded Control Stacks is now
  supported.

* New "--binary-output" command line option instructs GDB to set the
  translation mode of its stdout/stderr to binary mode.  This disables
  Line Feed translation.  MS-Windows only.

* GDB now adds all type symbols to the .gdb_index section.  The index
  version number has not increased as a consequence of this change.
  This fixes an issue where GDB could fail to find a type when relying
  on the index.  Any existing indexes should be regenerated.

* New commands

maintenance check psymtabs
  Renamed from maintenance check-psymtabs

maintenance check symtabs
  Renamed from maintenance check-symtabs

maintenance canonicalize
  Show the canonical form of a C++ name.

set riscv numeric-register-names on|off
show riscv numeric-register-names
  Controls whether GDB refers to risc-v registers by their numeric names
  (e.g 'x1') or their abi names (e.g. 'ra').
  Defaults to 'off', matching the old behaviour (abi names).

set style emoji on|off|auto
show style emoji
  Controls whether GDB can display emoji.  The default is "auto",
  which means emoji will be displayed in some situations when
  the host charset is UTF-8.

set style warning-prefix STRING
set style error-prefix STRING
  These commands control the prefix that is printed before warnings
  and errors, respectively.  This functionality is intended for use
  with emoji display, and so the prefixes are only displayed if emoji
  styling is enabled.

info linker-namespaces
info linker-namespaces [[N]]
  Print information about the given linker namespace (identified as N),
  or about all the namespaces if no argument is given.

* Changed commands

info sharedlibrary
  On Linux and FreeBSD, the addresses shown in the output of this
  command are now for the full memory range allocated to the shared
  library.

info threads [-gid] [-stopped] [-running] [ID]...
  If no threads match the given ID(s) or filter options, GDB now prints

    No threads matched.

  without printing the provided arguments.  The newly added '-stopped'
  option makes GDB list the stopped threads only.  Similarly,
  '-running' makes GDB list the running threads only.  If both options
  are given together, both stopped and running threads are listed.
  These new flags can be useful to get a reduced list when there is a
  large number of threads.

* GDB-internal Thread Local Storage (TLS) support

  ** Linux targets for the x86_64, aarch64, ppc64, s390x, and riscv
     architectures now have GDB-internal support for TLS address
     lookup in addition to that traditionally provided by the
     libthread_db library.  This internal support works for programs
     linked against either the GLIBC or MUSL C libraries.  For
     programs linked against MUSL, this new internal support provides
     new debug functionality, allowing access to TLS variables, due to
     the fact that MUSL does not implement the libthread_db library.
     Internal TLS support is also useful in cross-debugging
     situations, debugging statically linked binaries, and debugging
     programs linked against GLIBC 2.33 and earlier, but which are not
     linked against libpthread.

  ** The command 'maint set force-internal-tls-address-lookup on' may
     be used to force the internal TLS lookup mechanisms to be used.
     Otherwise, TLS lookup via libthread_db will still be preferred,
     when available.

* Python API

  ** GDB no longer supports Python versions less than 3.4.

  ** New class gdb.Color for dealing with colors.

  ** New constant gdb.PARAM_COLOR represents color type of a
     gdb.Parameter.value.  Parameter's value is gdb.Color instance.

  ** The memory_source argument (the second argument) has been removed

gdb/contrib/gdb-add-index.sh
#! /bin/sh

# Add a .gdb_index section to a file.

# Copyright (C) 2010-2024 Free Software Foundation, Inc.
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# This program assumes gdb and objcopy are in $PATH.
# If not, or you want others, pass the following in the environment
GDB=${GDB:=gdb}
OBJCOPY=${OBJCOPY:=objcopy}
READELF=${READELF:=readelf}

myname="${0##*/}"

dwarf5=""
if [ "$1" = "-dwarf-5" ]; then
    dwarf5="$1"
    shift
fi

if test $# != 1; then
    echo "usage: $myname [-dwarf-5] FILE" 1>&2
    exit 1
fi

file="$1"

if test -L "$file"; then
    target=$(readlink "$file")
    st=$?
    if [ $st -eq 127 ]; then
	echo "$myname: 'readlink' missing.  Failed to follow symlink $1." 1>&2
	exit 1
    fi

    # Count number of links followed in order to detect loops.
    count=0
    while test -L "$file"; do
	target=$(readlink "$file")

	case "$target" in
	    /*)
		file="$target"
		;;
	    *)
		file="$(dirname "$file")/$target"
		;;
	esac

	count="$((count + 1))"
	if test "$count" -gt 10; then
	    echo "$myname: Detected loop while following link $file"
	    exit 1
	fi
    done
fi

if test ! -r "$file"; then
    echo "$myname: unable to access: $file" 1>&2
    exit 1
fi

dir="${file%/*}"
test "$dir" = "$file" && dir="."

dwz_file=""
if $READELF -S "$file" | grep -q " \.gnu_debugaltlink "; then
    dwz_file=$($READELF --string-dump=.gnu_debugaltlink "$file" \
		   | grep -A1  "'\.gnu_debugaltlink':" \
		   | tail -n +2 \
		   | sed 's/.*]//')
    dwz_file=$(echo $dwz_file)
    if $READELF -S "$dwz_file" | grep -E -q " \.(gdb_index|debug_names) "; then
	# Already has an index, skip it.
	dwz_file=""
    fi
fi

set_files ()
{
    fpath="$1"

    index4="${fpath}.gdb-index"
    index5="${fpath}.debug_names"
    debugstr="${fpath}.debug_str"
    debugstrmerge="${fpath}.debug_str.merge"
    debugstrerr="${fpath}.debug_str.err"
}

tmp_files=
for f in "$file" "$dwz_file"; do
    if [ "$f" = "" ]; then
	continue
    fi
    set_files "$f"
    tmp_files="$tmp_files $index4 $index5 $debugstr $debugstrmerge $debugstrerr"
done

rm -f $tmp_files

# Ensure intermediate index file is removed when we exit.
trap "rm -f $tmp_files" 0

$GDB --batch -nx -iex 'set auto-load no' \
    -iex 'set debuginfod enabled off' \
    -ex "file '$file'" -ex "save gdb-index $dwarf5 '$dir'" || {
    # Just in case.
    status=$?
    echo "$myname: gdb error generating index for $file" 1>&2
    exit $status
}

# In some situations gdb can exit without creating an index.  This is
# not an error.
# E.g., if $file is stripped.  This behavior is akin to stripping an
# already stripped binary, it's a no-op.
status=0

handle_file ()
{
    fpath="$1"

    set_files "$fpath"

    if test -f "$index4" -a -f "$index5"; then
	echo "$myname: Both index types were created for $fpath" 1>&2
	status=1
    elif test -f "$index4" -o -f "$index5"; then
	if test -f "$index4"; then
	    index="$index4"
	    section=".gdb_index"
	else
	    index="$index5"
	    section=".debug_names"
	fi
	if test -s "$debugstr"; then
	    if ! $OBJCOPY --dump-section .debug_str="$debugstrmerge" "$fpath" \
		 /dev/null 2> "$debugstrerr"; then
		cat >&2 "$debugstrerr"
		exit 1
	    fi
	    cat "$debugstr" >>"$debugstrmerge"
	    if grep -q "can't dump section '.debug_str' - it does not exist" \
		    "$debugstrerr"; then
		$OBJCOPY --add-section $section="$index" \
			 --set-section-flags $section=readonly \
			 --add-section .debug_str="$debugstrmerge" \
		         --set-section-flags .debug_str=readonly \
			 "$fpath" "$fpath"
	    else
		$OBJCOPY --add-section $section="$index" \
			 --set-section-flags $section=readonly \
			 --update-section .debug_str="$debugstrmerge" \
			 "$fpath" "$fpath"
	    fi
	else
	    $OBJCOPY --add-section $section="$index" \
		     --set-section-flags $section=readonly \
		     "$fpath" "$fpath"
	fi

	status=$?
    else
	echo "$myname: No index was created for $fpath" 1>&2
	echo "$myname: [Was there no debuginfo? Was there already an index?]" \
	     1>&2
    fi
}

handle_file "$file"
if [ "$dwz_file" != "" ]; then
    handle_file "$dwz_file"
fi

exit $status
